File: src/main/java/com/example/lee/thirdRound/FirstMissingPositive.java
package com.example.lee.thirdRound;

import java.util.Optional;
import java.util.stream.IntStream;

/**
 * Created by benbendaisy on 7/6/17.
 */
public class FirstMissingPositive {
    public int firstMissingPositive(int[] nums) {
        if (nums == null || nums.length < 1) {
            return 1;
        } else if (nums.length == 1) {
            return nums[0] == 1 ? 2 : 1;
        }
        for (int i = 0; i < nums.length; i++) {
            while (nums[i] != i && nums[i] < nums.length && nums[i] > 0 && nums[nums[i]] != nums[i]) {
                swap(nums, i, nums[i]);
            }
        }
        Optional<Integer> missingIntegerOpt = IntStream.range(1, nums.length)
                .boxed()
                .filter(i -> nums[i] != i)
                .findAny();
        return missingIntegerOpt.isPresent() ?
                missingIntegerOpt.get() : (nums[0] == nums.length ? nums.length + 1 : nums.length);
    }

    private void swap(int[] nums, int i, int j) {
        int t = nums[i];
        nums[i] = nums[j];
        nums[j] = t;
    }

    public static void main(String[] args) {
        FirstMissingPositive firstMissingPositive = new FirstMissingPositive();
        int[] nums = {1, 0};
        System.out.println(firstMissingPositive.firstMissingPositive(nums));
    }
}

File: python_module/examples/first_missing_positive.py
from typing import List


class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        if not nums:
            return 1

        for i in range(len(nums)):
            while nums[i] != i and nums[i] < len(nums) and nums[i] > 0 and nums[nums[i]] != nums[i]:
                # nums[i], nums[nums[i]] = nums[nums[i]], nums[i]
                nums[nums[i]], nums[i] = nums[i], nums[nums[i]]

        for i in range(1, len(nums)):
            if nums[i] != i:
                return i

        return len(nums) + 1 if nums[0] == len(nums) else len(nums)

if __name__ == "__main__":
    nums = [1,2,0]
    solution = Solution()
    ret = solution.firstMissingPositive(nums)
    print(ret)